File: Structural/Adapter_C++/Adapter1.cpp
// Adapter Design Pattern - Structural Category

// An Adapter Design Pattern converts the interface of a class into another interface the clients expect. 
// Adapter lets classes work together that could not otherwise because of incompatible interfaces.

// An Adapter pattern comprises three components:
// Target: This is the interface with which the clients interact.
// Adaptee: This is the interface the client wants to interact with, but cannot without the help of the Adapter.
// Adapter: This is derived from Target and contains the object of Adaptee.

// http://www.codeproject.com/Tips/595716/Adapter-Design-Pattern-in-Cplusplus

#include <iostream>

using namespace std;

// Abstract Target: This is the interface the client interacts with.
class AbstractPlug 
{
	public:

		void virtual RoundPin() { }
		void virtual PinCount() { }
};

// Concrete Target
class Plug : public AbstractPlug 
{
	public:

		void RoundPin() { cout << "I am Round Pin" << endl; }
		void PinCount() { cout << "I have two pins" << endl; }
};


// Abstract Adaptee: This is the interface the client wants to interact with, 
// but cannot interact without the help of the Adapter.
class AbstractSwitchBoard 
{
	public:

		void virtual FlatPin() { }
		void virtual PinCount() { }
};

// Concrete Adaptee
class SwitchBoard : public AbstractSwitchBoard 
{
	public:

		void FlatPin() { cout << "I am Flat Pin" << endl; }
		void PinCount() { cout << "I have three pins" << endl; }
};

// Adapter: This is derived from Target and contains the object of Adaptee.
class Adapter : public AbstractPlug 
{
	private:

		AbstractSwitchBoard * _asb;

	public:

		Adapter(AbstractSwitchBoard * asb) { _asb = asb; }

		void RoundPin() { _asb->FlatPin(); }
		void PinCount() { _asb->PinCount(); }
};

// Client code
int main() 
{
	// Adaptee
	SwitchBoard *mySwitchBoard = new SwitchBoard; 
	
	// Target = Adapter(Adaptee)
	AbstractPlug * adapter = new Adapter(mySwitchBoard);

	adapter->RoundPin();
	adapter->PinCount();

	system("pause");

	return 0;
}

File: Structural/Adapter_C#/Adapter1.cs
// Adapter Design Patter - Structural Category

// Convert the interface of a class into another interface clients expect. 
// Adapter lets classes work together that could not otherwise because of incompatible interfaces.

// This structural code demonstrates the Adapter design pattern which maps 
// the interface of one class onto another so that they can work together. 
// These incompatible classes may come from different libraries or frameworks.

// http://www.dofactory.com/net/adapter-design-pattern

using System;

namespace Adapter
{
    /// <summary>
    /// MainApp startup class for Structural
    /// Adapter Design Pattern.
    /// </summary>
    class MainApp
    {
        /// <summary>
        /// Entry point into console application.
        /// </summary>
        static void Main()
        {
            // Create adapter and place a request
            Target target = new Adapter();
            target.Request();

            // Wait for user
            Console.ReadKey();
        }
    }

    /// <summary>
    /// The 'Target' class
    /// </summary>
    class Target
    {
        public virtual void Request()
        {
            Console.WriteLine("Called Target Request()");
        }
    }

    /// <summary>
    /// The 'Adapter' class
    /// </summary>
    class Adapter : Target
    {
        private Adaptee _adaptee = new Adaptee();

        public override void Request()
        {
            // Possibly do some other work and then call SpecificRequest
            _adaptee.SpecificRequest();
        }
    }

    /// <summary>
    /// The 'Adaptee' class
    /// </summary>
    class Adaptee
    {
        public void SpecificRequest()
        {
            Console.WriteLine("Called SpecificRequest()");
        }
    }
}